pkg/controllers/terminating/v1alpha1/controller.go

```diff
@@ -63,22 +65,68 @@ func NewController(kubeClient client.Client, coreV1Client corev1.CoreV1Interface
 	}
 }
 
-// Reconcile executes a reallocation control loop for the resource
+// Reconcile executes a termination control loop for the resource
+// Requeue a reconcile.Request if:
+//   1) Cordon, Drain, or Terminate fails
+//   2) Drain succeeds, but node has non-daemonset pods
 func (c *Controller) Reconcile(ctx context.Context, object client.Object) (reconcile.Result, error) {
-	// 1. Cordon terminable nodes
-	if err := c.terminator.cordonNodes(ctx); err != nil {
-		return reconcile.Result{}, fmt.Errorf("cordoning terminable nodes, %w", err)
+	node := object.(*v1.Node)
+	// 0. Don't do work on a non-terminable node
+	if phase, ok := node.Labels[provisioning.ProvisionerPhaseLabel]; !ok || phase == provisioning.ProvisionerUnderutilizedPhase {
+		return reconcile.Result{}, nil
 	}
 
-	// 2. Drain and delete nodes
-	if err := c.terminator.terminateNodes(ctx); err != nil {
-		return reconcile.Result{}, fmt.Errorf("terminating nodes, %w", err)
+	// 1. Cordon terminable nodes
+	if err := c.terminator.cordonNode(ctx, node); err != nil {
+		return reconcile.Result{Requeue: true}, fmt.Errorf("cordoning terminable nodes, %w", err)
+	}
+	// 2. Drain node
+	drained, err := c.terminator.drainNode(ctx, node)
+	if err != nil {
+		return reconcile.Result{Requeue: true}, fmt.Errorf("draining nodes, %w", err)
+	}
+	// 3. Terminate and Delete node
+	if drained {
+		if err := c.terminator.terminateNode(ctx, node); err != nil {
+			return reconcile.Result{Requeue: true}, fmt.Errorf("terminating nodes, %w", err)
+		}
 	}
-	return reconcile.Result{}, nil
+	// 4. Requeue if we still need to drain
+	return reconcile.Result{Requeue: !drained}, nil
 }
 
-func (c *Controller) Watches(context.Context) (source.Source, handler.EventHandler, builder.WatchesOption) {
+func (c *Controller) Watches() (source.Source, handler.EventHandler, builder.WatchesOption) {
 	return &source.Kind{Type: &v1.Node{}},
 		&handler.EnqueueRequestForObject{},
-		builder.WithPredicates(predicate.NewPredicateFuncs(func(object client.Object) bool { return false }))
+		builder.WithPredicates(
+			predicate.Funcs{
+				CreateFunc: func(e event.CreateEvent) bool {
+					return false
+				},
+				// kubectl delete node is not supported yet
+				DeleteFunc: func(e event.DeleteEvent) bool {
+					return false
+				},
+				UpdateFunc: func(e event.UpdateEvent) bool {
+					nodeLabels := e.ObjectNew.GetLabels()
+					// Don't watch nodes not created by provisioners
+					if _, ok := nodeLabels[provisioning.ProvisionerNameLabelKey]; !ok {
+						return false
+					}
+					if _, ok := nodeLabels[provisioning.ProvisionerNamespaceLabelKey]; !ok {
+						return false
+					}
+					if phase, ok := nodeLabels[provisioning.ProvisionerPhaseLabel]; ok {
+						return (phase == provisioning.ProvisionerTerminablePhase ||
+							phase == provisioning.ProvisionerDrainingPhase ||
+							phase == provisioning.ProvisionerDrainedPhase) &&
+							e.ObjectOld.GetResourceVersion() != e.ObjectNew.GetResourceVersion()
```

Comment:
Why do we need to check resource versions?

Reply:
We do this check to make sure we don't take action on a cache resync event. When we resync the watch client cache, we send an update, but the resource versions will be the same, as we are not changing the object itself.

https://github.com/kubernetes/kubernetes/blob/ea0764452222146c47ec826977f49d7001b0ea8c/staging/src/k8s.io/client-go/tools/cache/shared_informer.go#L152

Reply:
The point of the resync interval is to trigger the reconciliation logic in case something has accidentally gotten stale.